File: array_sort/sort_merge/src/MergeSortDemo.cpp
// C++
// 配列の並び替え: マージソート (Merge Sort)

#include <iostream>
#include <vector>

class ArrayData {
private:
    std::vector<int> _data;

    std::vector<int> _merge_sort(const std::vector<int>& target) {
        // 配列の長さが1以下の場合はそのまま返す（基本ケース）
        if (target.size() <= 1) {
            return target;
        }
        
        // 配列を半分に分割
        size_t mid = target.size() / 2;
        std::vector<int> left_half(target.begin(), target.begin() + mid);
        std::vector<int> right_half(target.begin() + mid, target.end());
        
        // 左右の半分を再帰的にソート
        left_half = _merge_sort(left_half);
        right_half = _merge_sort(right_half);
        
        // ソート済みの半分同士をマージ
        return _merge(left_half, right_half);
    }

    std::vector<int> _merge(const std::vector<int>& left, const std::vector<int>& right) {
        std::vector<int> result;
        size_t i = 0, j = 0;
        
        // 左右の配列を比較しながらマージ
        while (i < left.size() && j < right.size()) {
            if (left[i] <= right[j]) {
                result.push_back(left[i]);
                i++;
            } else {
                result.push_back(right[j]);
                j++;
            }
        }
        
        // 残った要素を追加
        while (i < left.size()) {
            result.push_back(left[i]);
            i++;
        }
        
        while (j < right.size()) {
            result.push_back(right[j]);
            j++;
        }
        
        return result;
    }

public:
    std::vector<int> get() {
        return _data;
    }

    bool set(const std::vector<int>& data) {
        _data = data;
        return true;
    }

    bool sort() {
        _data = _merge_sort(_data);
        return true;
    }
};

void printVector(const std::vector<int>& vec) {
    std::cout << "  [";
    for (size_t i = 0; i < vec.size(); ++i) {
        std::cout << vec[i];
        if (i < vec.size() - 1) {
            std::cout << ", ";
        }
    }
    std::cout << "]" << std::endl;
}

int main() {
    std::cout << "MergeSort TEST -----> start" << std::endl;

    ArrayData array_data;

    // ランダムな整数の配列
    std::cout << "\nsort" << std::endl;
    std::vector<int> input1 = {64, 34, 25, 12, 22, 11, 90};
    std::cout << "  ソート前: ";
    printVector(input1);
    array_data.set(input1);
    array_data.sort();
    std::cout << "  ソート後: ";
    printVector(array_data.get());
    
    // 既にソートされている配列
    std::cout << "\nsort" << std::endl;
    std::vector<int> input2 = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
    std::cout << "  ソート前: ";
    printVector(input2);
    array_data.set(input2);
    array_data.sort();
    std::cout << "  ソート後: ";
    printVector(array_data.get());
    
    // 逆順の配列
    std::cout << "\nsort" << std::endl;
    std::vector<int> input3 = {10, 9, 8, 7, 6, 5, 4, 3, 2, 1};
    std::cout << "  ソート前: ";
    printVector(input3);
    array_data.set(input3);
    array_data.sort();
    std::cout << "  ソート後: ";
    printVector(array_data.get());
    
    // 重複要素を含む配列
    std::cout << "\nsort" << std::endl;
    std::vector<int> input4 = {10, 9, 8, 7, 6, 10, 9, 8, 7, 6};
    std::cout << "  ソート前: ";
    printVector(input4);
    array_data.set(input4);
    array_data.sort();
    std::cout << "  ソート後: ";
    printVector(array_data.get());
    
    // 空の配列
    std::cout << "\nsort" << std::endl;
    std::vector<int> input5 = {};
    std::cout << "  ソート前: ";
    printVector(input5);
    array_data.set(input5);
    array_data.sort();
    std::cout << "  ソート後: ";
    printVector(array_data.get());

    std::cout << "\nMergeSort TEST <----- end" << std::endl;
    
    return 0;
}

File: array_sort/sort_merge/src/MergeSortDemo.rs
// Rust
// 配列の並び替え: マージソート (Merge Sort)

struct ArrayData {
    data: Vec<i32>,
}

impl ArrayData {
    fn new() -> Self {
        ArrayData { data: Vec::new() }
    }

    fn get(&self) -> &Vec<i32> {
        &self.data
    }

    fn set(&mut self, data: Vec<i32>) -> bool {
        self.data = data;
        true
    }

    fn merge_sort(&mut self) -> bool {
        let sorted = self._merge_sort(self.data.clone());
        self.data = sorted;
        true
    }

    fn _merge_sort(&self, target: Vec<i32>) -> Vec<i32> {
        // 配列の長さが1以下の場合はそのまま返す（基本ケース）
        if target.len() <= 1 {
            return target;
        }
        
        // 配列を半分に分割
        let mid = target.len() / 2;
        let left_half = target[..mid].to_vec();
        let right_half = target[mid..].to_vec();
        
        // 左右の半分を再帰的にソート
        let left_half = self._merge_sort(left_half);
        let right_half = self._merge_sort(right_half);
        
        // ソート済みの半分同士をマージ
        self._merge(left_half, right_half)
    }

    fn _merge(&self, left: Vec<i32>, right: Vec<i32>) -> Vec<i32> {
        let mut result = Vec::new();
        let mut i = 0;
        let mut j = 0;
        
        // 左右の配列を比較しながらマージ
        while i < left.len() && j < right.len() {
            if left[i] <= right[j] {
                result.push(left[i]);
                i += 1;
            } else {
                result.push(right[j]);
                j += 1;
            }
        }
        
        // 残った要素を追加
        result.extend_from_slice(&left[i..]);
        result.extend_from_slice(&right[j..]);
        
        result
    }
}

fn main() {
    println!("MergeSort TEST -----> start");

    let mut array_data = ArrayData::new();

    // ランダムな整数の配列
    println!("\nsort");
    let input = vec![64, 34, 25, 12, 22, 11, 90];
    println!("  ソート前: {:?}", input);
    array_data.set(input);
    array_data.merge_sort();
    println!("  ソート後: {:?}", array_data.get());
    
    // 既にソートされている配列
    println!("\nsort");
    let input = vec![1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
    println!("  ソート前: {:?}", input);
    array_data.set(input);
    array_data.merge_sort();
    println!("  ソート後: {:?}", array_data.get());
    
    // 逆順の配列
    println!("\nsort");
    let input = vec![10, 9, 8, 7, 6, 5, 4, 3, 2, 1];
    println!("  ソート前: {:?}", input);
    array_data.set(input);
    array_data.merge_sort();
    println!("  ソート後: {:?}", array_data.get());
    
    // 重複要素を含む配列
    println!("\nsort");
    let input = vec![10, 9, 8, 7, 6, 10, 9, 8, 7, 6];
    println!("  ソート前: {:?}", input);
    array_data.set(input);
    array_data.merge_sort();
    println!("  ソート後: {:?}", array_data.get());
    
    // 空の配列
    println!("\nsort");
    let input: Vec<i32> = vec![];
    println!("  ソート前: {:?}", input);
    array_data.set(input);
    array_data.merge_sort();
    println!("  ソート後: {:?}", array_data.get());

    println!("\nMergeSort TEST <----- end");
}